### ISSUE TITLE
paddle 训练DCGAN生成Mnist，相同的逻辑在1.8版本跑出效果了，改到2.0跑不出效果，找不出问题出在哪儿了，求助！！！

### ISSUE THREAD

**Issue by KobeBryant-0824** [PaddlePaddle/Paddle#31623]
2.0代码如下：
文件一：basic_dataloader.py 文件数据加载
import numpy as np
import paddle
# 由于paddle.io.DataLoader只接受numpy ndarray或者paddle Tensor作为数据输入
# 该自定义类将MNIST数据reshape并转化为numpy ndarray类型，并且将数据从[0, 255] 转化到 [-1, 1]
class MnistDataset(paddle.vision.datasets.MNIST):
    def __init__(self, mode, return_label=True):
        super(MnistDataset, self).__init__(mode=mode)
        self.return_label = return_label

    def __getitem__(self, idx):
        img = np.reshape(self.images[idx], [1, 28, 28])
        img = img / 255.0 * 2.0 - 1.0
        if self.return_label:
            return img, np.array(self.labels[idx]).astype('int64')
        return img,

    def __len__(self):
        return len(self.images)
文件二：basic_model.py 定义网络结构---生成器、判别器
import numpy as np
import paddle

class Up(paddle.nn.Layer):
    def __init__(self,  in_channels, 
                        out_channels,
                        kernel_size,
                        stride,
                        padding,
                        bias_attr=None,
                        weight_attr=None):
        super(Up, self).__init__()
        
        self._conv2DTranspose = paddle.nn.Conv2DTranspose(
                                    in_channels = in_channels,
                                    out_channels = out_channels,
                                    kernel_size =  kernel_size,
                                    stride = stride,
                                    padding = padding,
                                    bias_attr=bias_attr, 
                                    weight_attr=weight_attr)
        self._batchNorm2D = paddle.nn.BatchNorm2D(
                                    num_features = out_channels)
        self._reLU = paddle.nn.ReLU()
        
    def forward(self, inputs):
        x = self._conv2DTranspose(inputs)
        x = self._batchNorm2D(x)
        x = self._reLU(x)
        return x

class Down(paddle.nn.Layer):
    def __init__(self, in_channels, 
                        out_channels,
                        kernel_size,
                        stride,
                        padding,
                        bias_attr=None,
                        weight_attr=None):
        super(Down, self).__init__()
        
        self._conv2D = paddle.nn.Conv2D(
                                    in_channels = in_channels,
                                    out_channels = out_channels, 
                                    kernel_size = kernel_size,
                                    stride = stride,
                                    padding = padding,
                                    bias_attr=bias_attr, 
                                    weight_attr=weight_attr)
        self._batchNorm2D = paddle.nn.BatchNorm2D(
                                    num_features = out_channels)
        self._leakyReLU = paddle.nn.LeakyReLU(negative_slope=0.2)
    
    def forward(self, inputs):
        x = self._conv2D(inputs)
        x = self._batchNorm2D(x)
        x = self._leakyReLU(x)

        return x

class Generator(paddle.nn.Layer):
    def __init__(self, z_dim, channels_img, features_g, weight_attr=paddle.nn.initializer.Normal(0., 0.02)):
        super(Generator, self).__init__()
         
        self.up1 = Up(z_dim, features_g*16, 4, 1, 0, weight_attr = weight_attr, bias_attr=False)
        self.up2 = Up(features_g*16, features_g*8, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self.up3 = Up(features_g*8, features_g*4, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self.up4 = Up(features_g*4, features_g*2, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self._conv2DTranspose = paddle.nn.Conv2DTranspose(
                features_g*2, 
                channels_img, 
                kernel_size = 4, 
                stride = 2, 
                padding = 1,
                weight_attr = weight_attr,
                bias_attr=False
            )
        self._tanh = paddle.nn.Tanh()
        
    def forward(self, z):
        x = self.up1(z)
        x = self.up2(x)
        x = self.up3(x)
        x = self.up4(x)
        x = self._conv2DTranspose(x)
        x = self._tanh(x)

        return x

class Discriminator(paddle.nn.Layer):
    def __init__(self,channels_img, features_d, weight_attr=paddle.nn.initializer.Normal(0., 0.02)):
        super(Discriminator, self).__init__()
        
        self._conv2D1 = paddle.nn.Conv2D(
                channels_img, 
                features_d, 
                kernel_size = 4, 
                stride = 2, 
                padding = 1,
                weight_attr = weight_attr,
                bias_attr=False
            )
        self._reLU = paddle.nn.ReLU()
        self.down1 = Down(features_d, features_d*2, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self.down2 = Down(features_d*2, features_d*4, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self.down3 = Down(features_d*4, features_d*8, 4, 2, 1, weight_attr = weight_attr, bias_attr=False)
        self._conv2D2 = paddle.nn.Conv2D(
                features_d*8, 
                1, 
                kernel_size = 4,
                stride  = 2,
                padding = 0,
                weight_attr = weight_attr,
                bias_attr=False
            )
        self._sigmoid = paddle.nn.Sigmoid()
        
    def forward(self, image):
        x = self._conv2D1(image)
        x = self._reLU(x)
        x = self.down1(x)
        x = self.down2(x)
        x = self.down3(x)
        x = self._conv2D2(x)
        x = self._sigmoid(x)

        return x
   文件三：train.py 训练并保存模型，中间的损失值和fake结果保存到日志文件，用visualDL查看
import numpy as np
import paddle
import paddle.fluid as fluid
from basic_dataloader import MnistDataset
from basic_model import Generator, Discriminator
from visualdl import LogWriter

IMAGE_SIZE = 64
CHANNELS_IMAG = 1
FEATURES_DISC = 64
FEATURES_GEN = 64

Z_DIM = 100
BATCH_SIZE = 128
LEARNING_RATE = 1E-3

# 噪声生成，通过由噪声来生成假的图片数据输入。
def z_reader():
    while True:
        yield np.random.normal(0.0, 1.0, (Z_DIM, 1, 1)).astype('float32')
        
# 拼接一个batch图像用于VisualDL可视化
def concatenate_img(input_img):
    img_arr_broadcasted = ((np.zeros([BATCH_SIZE,3,28,28]) + input_img) * 255).astype('uint8').transpose((0,2,3,1)).reshape([-1,16,28,28,3])
    img_concatenated = np.concatenate(tuple(img_arr_broadcasted), axis=1)
    img_concatenated = np.concatenate(tuple(img_concatenated), axis=1)
    return img_concatenated
    
def train(train_loader, z_generator, epoch_num=10, batch_size=128, load_model=False):
    # 模型存储路径
    model_path = './output/'
    bce_loss = paddle.nn.BCELoss()
    # 定义判别器与生成器对象
    d = Discriminator(CHANNELS_IMAG,FEATURES_DISC)
    d.train()
    g = Generator(Z_DIM, CHANNELS_IMAG, FEATURES_GEN)
    g.train()

    # 创建优化方法
    g_optimizer = paddle.optimizer.Adam(learning_rate=LEARNING_RATE, beta1=0.5, beta2=0.999,  parameters=g.parameters())
    d_optimizer = paddle.optimizer.Adam(learning_rate=LEARNING_RATE, beta1=0.5, beta2=0.999, parameters=d.parameters())
        
    # 定义日志写入(先清空日志文件夹)
    d_loss_wrt = LogWriter(logdir='./log/d_loss')
    g_loss_wrt = LogWriter(logdir='./log/g_loss')
    image_wrt = LogWriter(logdir='./log/imgs')

    iteration_num = 0
    for epoch in range(epoch_num):
        for i, data in enumerate(train_loader()):

            real_image = data[0]
            # 丢弃不满整个batch_size的数据
            if(len(real_image) != BATCH_SIZE):
                continue
            iteration_num += 1
            
            '''
            判别器d通过最小化输入真实图片时判别器d的输出与真值标签ones的交叉熵损失，来优化判别器的参数，
            以增加判别器d识别真实图片real_image为真值标签ones的概率。
            '''
            # 将MNIST数据集里的图片读入real_image，将真值标签ones用数字1初始化
            real_image = paddle.to_tensor(real_image)
            real_image = fluid.layers.image_resize(real_image, out_shape=[IMAGE_SIZE,IMAGE_SIZE])
            ones = paddle.to_tensor(np.ones([len(real_image), 1]).astype('float32'))
            # 计算判别器d判断真实图片的概率
            p_real = d(real_image)
            # 计算判别真图片为真的损失
            p_real = paddle.reshape(p_real, shape=[len(real_image), -1])
            #print(p_real.min(),p_real.max())
            real_cost = bce_loss(p_real, ones)
            #print('real_cost.shape',real_cost.shape)

            '''
            判别器d通过最小化输入生成器g生成的假图片g(z)时判别器的输出与假值标签zeros的交叉熵损失，
            来优化判别器d的参数，以增加判别器d识别生成器g生成的假图片g(z)为假值标签zeros的概率。
            '''
            # 创建高斯分布的噪声z，将假值标签zeros初始化为0
            z = next(z_generator())
            z = paddle.to_tensor(np.array(z))
            zeros = paddle.to_tensor(np.zeros([len(real_image), 1]).astype('float32'))
            # 判别器d判断生成器g生成的假图片的概率
            p_fake = d(g(z))
            # 计算判别生成器g生成的假图片为假的损失
            p_fake = paddle.reshape(p_fake, shape=[len(real_image), -1])
            #print(p_fake.min(),p_fake.max())
            fake_cost = bce_loss(p_fake, zeros)
            #print('fake_cost.shape',fake_cost.shape)

            # 反向传播更新判别器d的参数
            d_loss = real_cost + fake_cost
            d_loss.backward()
            d_optimizer.minimize(d_loss)
            d_optimizer.step()
            d.clear_gradients()

            '''
            生成器g通过最小化判别器d判别生成器生成的假图片g(z)为真的概率d(fake)与真值标签ones的交叉熵损失，
            来优化生成器g的参数，以增加生成器g使判别器d判别其生成的假图片g(z)为真值标签ones的概率。
            '''
            # 生成器用输入的高斯噪声z生成假图片
            fake = g(z)
            # 计算判别器d判断生成器g生成的假图片的概率
            p_confused = d(fake)
            # 使用判别器d判断生成器g生成的假图片的概率与真值ones的交叉熵计算损失
            p_confused = paddle.reshape(p_confused, shape=[len(real_image), -1])
            #print(p_confused.min(),p_confused.max())
            g_cost = bce_loss(p_confused, ones)
            #print('g_cost.shape',g_cost.shape)

            # 反向传播更新生成器g的参数
            g_cost.backward()
            g_optimizer.minimize(g_cost)
            g_optimizer.step()
            g.clear_gradients()
            
            # 打印输出  
            print('epoch =', epoch, ', batch =', i, 
                ', real_d_loss =', real_cost.numpy(), 
                ', fake_d_loss =', fake_cost.numpy(), 
                ', g_loss =', g_cost.numpy())
            
            # 写VisualDL日志
            d_loss_wrt.add_scalar(tag='loss', step=iteration_num, value=d_loss.numpy())
            g_loss_wrt.add_scalar(tag='loss', step=iteration_num, value=g_cost.numpy())
            fake = fluid.layers.image_resize(fake, out_shape=[28,28])
            image_wrt.add_image(tag='numbers', img=concatenate_img(fake.numpy()), step=iteration_num)
    
    # 存储模型
    paddle.save(g.state_dict(), model_path+'g.pdparams')
    paddle.save(d.state_dict(), model_path+'d.pdparams')

        
def main():

    train_loader = paddle.io.DataLoader(
        MnistDataset(mode='train'), batch_size=BATCH_SIZE, drop_last=True)
    
    # 生成假图片的reader
    z_generator = paddle.batch(z_reader, batch_size=BATCH_SIZE)
    
    train(train_loader, z_generator, epoch_num=10, batch_size=BATCH_SIZE)

if __name__ == "__main__":
    main()

**Comment by KobeBryant-0824** [comment]
问题已经解决了
